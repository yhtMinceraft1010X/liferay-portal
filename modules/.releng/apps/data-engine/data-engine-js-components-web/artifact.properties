artifact.git.id=b80ee6bc52c0532e058704f1fefbd2d16f6d6e2a
artifact.jsdoc.url=https://repository-cdn.liferay.com/nexus/content/repositories/liferay-public-releases/com/liferay/com.liferay.data.engine.js.components.web/1.0.60/com.liferay.data.engine.js.components.web-1.0.60-jsdoc.jar
artifact.sources-commercial.url=https://repository-cdn.liferay.com/nexus/content/repositories/liferay-public-releases/com/liferay/com.liferay.data.engine.js.components.web/1.0.60/com.liferay.data.engine.js.components.web-1.0.60-sources-commercial.jar
artifact.sources.url=https://repository-cdn.liferay.com/nexus/content/repositories/liferay-public-releases/com/liferay/com.liferay.data.engine.js.components.web/1.0.60/com.liferay.data.engine.js.components.web-1.0.60-sources.jar
artifact.url=https://repository-cdn.liferay.com/nexus/content/repositories/liferay-public-releases/com/liferay/com.liferay.data.engine.js.components.web/1.0.60/com.liferay.data.engine.js.components.web-1.0.60.jar