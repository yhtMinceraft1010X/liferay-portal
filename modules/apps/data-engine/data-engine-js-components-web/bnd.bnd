Bundle-Name: Liferay Data Engine JS Components Web
Bundle-SymbolicName: com.liferay.data.engine.js.components.web
Bundle-Version: 1.0.61
Web-ContextPath: /data-engine-js-components-web