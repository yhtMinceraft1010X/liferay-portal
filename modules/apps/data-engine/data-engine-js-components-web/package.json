{
	"dependencies": {
		"dom-align": "1.10.4",
		"moment": "^2.22.2"
	},
	"jest": {
		"collectCoverageFrom": [
			"src/main/resources/META-INF/resources/js/**/*.es.js"
		],
		"modulePathIgnorePatterns": [
			"/__mock__/",
			"/__utils__/"
		],
		"setupFiles": [
			"<rootDir>/jest-setup.config.js"
		],
		"snapshotSerializers": [
			"metal-jest-serializer"
		],
		"testPathIgnorePatterns": [
			"<rootDir>/test/js/[^/]+\\.js"
		]
	},
	"main": "js/index.es.js",
	"name": "data-engine-js-components-web",
	"private": true,
	"scripts": {
		"build": "liferay-npm-scripts build",
		"checkFormat": "liferay-npm-scripts check",
		"format": "liferay-npm-scripts fix",
		"test": "metalsoy -s \"test/**/*.soy\" -d \"test\" && liferay-npm-scripts test"
	},
	"version": "1.0.61"
}
