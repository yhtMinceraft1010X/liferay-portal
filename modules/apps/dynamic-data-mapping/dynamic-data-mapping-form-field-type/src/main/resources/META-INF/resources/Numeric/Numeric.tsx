/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {ClayInput} from '@clayui/form';
import classNames from 'classnames';

// @ts-ignore

import {SettingsContext, useFormState} from 'data-engine-js-components-web';
import React, {ChangeEventHandler, FocusEventHandler, useMemo} from 'react';
import {createNumberMask} from 'text-mask-addons';
import {conformToMask} from 'text-mask-core';

import {FieldBase} from '../FieldBase/ReactFieldBase.es';
import {ISymbols} from '../NumericInputMask/NumericInputMask';
import {trimLeftZero} from '../util/numericalOperations';

// @ts-ignore

import withConfirmationField from '../util/withConfirmationField.es';

import './Numeric.scss';

import type {FieldChangeEventHandler, Locale, LocalizedValue} from '../types';

const NON_NUMERIC_REGEX = /[\D]/g;

const adaptiveMask = (rawValue: string, inputMaskFormat: string) => {
	const generateMask = (mask: string): string => {
		if (!mask.includes('0')) {
			return mask;
		}

		const inputNumbers = rawValue.match(/\d/g)?.length ?? 0;
		const mandatorySize = mask.match(/9/g)?.length ?? 0;
		if (inputNumbers <= mandatorySize) {
			return mask.replace(/0/g, '');
		}

		return generateMask(mask.replace('0', '9'));
	};

	return [...generateMask(inputMaskFormat)].map((char) =>
		char === '9' ? /\d/ : char
	);
};

const getMaskedValue = ({
	dataType,
	decimalPlaces,
	includeThousandsSeparator = false,
	inputMaskFormat,
	symbols,
	value,
}: {
	dataType: NumericDataType;
	decimalPlaces: number;
	includeThousandsSeparator?: boolean;
	inputMaskFormat: string;
	symbols: ISymbols;
	value: string;
}): IMaskedNumber => {
	let mask;
	if (dataType === 'double') {
		const config: INumberMaskConfig = {
			allowDecimal: true,
			allowLeadingZeroes: true,
			allowNegative: true,
			decimalLimit: decimalPlaces,
			decimalSymbol: symbols.decimalSymbol,
			includeThousandsSeparator,
			prefix: '',
			thousandsSeparatorSymbol: symbols.thousandsSeparator ?? '',
		};

		mask = createNumberMask(config);
	}
	else {
		mask = adaptiveMask(value, inputMaskFormat);
	}

	const {conformedValue: masked} = conformToMask(value, mask, {
		guide: false,
		keepCharPositions: false,
		placeholderChar: '\u2000',
	});

	const regex = new RegExp(
		dataType === 'double' ? `[^-${symbols.decimalSymbol}\\d]` : '[^\\d]',
		'g'
	);

	return {
		masked,
		placeholder:
			dataType === 'double'
				? `0${symbols.decimalSymbol}${'0'.repeat(decimalPlaces)}`
				: inputMaskFormat.replace(/\d/g, '_'),
		raw: masked.replace(regex, ''),
	};
};

const getFormattedValue = ({
	dataType,
	symbols,
	value,
}: {
	dataType: NumericDataType;
	symbols: ISymbols;
	value: string;
}) => {
	if (!value) {
		return {masked: '', raw: ''};
	}

	const config: INumberMaskConfig = {
		allowLeadingZeroes: true,
		allowNegative: true,
		includeThousandsSeparator: false,
		prefix: '',
	};

	if (dataType === 'double') {
		config.allowDecimal = true;
		config.decimalLimit = null;
		config.decimalSymbol = symbols.decimalSymbol;
	}
	const mask = createNumberMask(config);

	const {conformedValue: masked}: {conformedValue: string} = conformToMask(
		value,
		mask,
		{
			guide: false,
			keepCharPositions: false,
			placeholderChar: '\u2000',
		}
	);

	return {
		masked: dataType === 'double' ? value : masked,
		raw: masked,
	};
};

const Numeric: React.FC<IProps> = ({
	append,
	appendType,
	dataType = 'integer',
	decimalPlaces,
	defaultLanguageId,
	id,
	inputMask,
	inputMaskFormat,
	localizedSymbols: initialLocalizedSymbols,
	localizedValue,
	name,
	onBlur,
	onChange,
	onFocus,
	placeholder,
	predefinedValue,
	readOnly,
	settingsContext,
	symbols: symbolsProp = {decimalSymbol: '.'},
	value,
	...otherProps
}) => {
	const {editingLanguageId}: {editingLanguageId: Locale} = useFormState();

	const localizedSymbols = settingsContext
		? SettingsContext.getSettingsContextProperty(
				settingsContext,
				'predefinedValue',
				'localizedSymbols'
		  )
		: initialLocalizedSymbols;

	const symbols = useMemo<ISymbols>(() => {
		if (inputMask) {
			return {
				decimalSymbol: symbolsProp.decimalSymbol,
				thousandsSeparator:
					symbolsProp.thousandsSeparator === 'none'
						? null
						: symbolsProp.thousandsSeparator,
			};
		}

		return localizedSymbols?.[editingLanguageId] || symbolsProp;
	}, [editingLanguageId, inputMask, localizedSymbols, symbolsProp]);

	const inputValue = useMemo<IMaskedNumber>(() => {
		let newValue =
			((localizedValue?.[editingLanguageId] ??
				localizedValue?.[defaultLanguageId]) ||
				value) ??
			predefinedValue ??
			'';

		if (dataType === 'double') {
			const symbolsValue = newValue.match(/[^-\d]/g);

			newValue = symbolsValue
				? newValue.replace(symbolsValue[0], symbols.decimalSymbol)
				: newValue;
		}

		return inputMask
			? getMaskedValue({
					dataType,
					decimalPlaces,
					includeThousandsSeparator: Boolean(
						symbols.thousandsSeparator
					),
					inputMaskFormat: String(inputMaskFormat),
					symbols,
					value: newValue,
			  })
			: {
					...getFormattedValue({
						dataType,
						symbols,
						value: newValue,
					}),
					placeholder,
			  };
	}, [
		dataType,
		decimalPlaces,
		defaultLanguageId,
		editingLanguageId,
		inputMask,
		inputMaskFormat,
		localizedValue,
		placeholder,
		predefinedValue,
		symbols,
		value,
	]);

	const handleChange: ChangeEventHandler<HTMLInputElement> = ({
		target: {value},
	}) => {
		value =
			inputMask && dataType === 'integer'
				? value
				: trimLeftZero({
						decimalSymbol: symbols.decimalSymbol,
						thousandsSeparator: symbols.thousandsSeparator,
						value,
				  });

		// allows user to delete characters from the mask

		const inputValueRaw = inputValue.raw.replace(NON_NUMERIC_REGEX, '');
		const rawValue = value.replace(NON_NUMERIC_REGEX, '');

		if (
			inputValue.masked?.length > value.length &&
			(inputValueRaw?.length ?? 0) === rawValue.length
		) {
			value = inputValueRaw.slice(0, -1);
		}

		const {masked, raw} = inputMask
			? getMaskedValue({
					dataType,
					decimalPlaces,
					inputMaskFormat: String(inputMaskFormat),
					symbols,
					value,
			  })
			: getFormattedValue({dataType, symbols, value});

		if (masked !== inputValue.masked) {
			onChange({target: {value: raw}});
		}
	};

	const input = (
		<ClayInput
			className={classNames({
				'ddm-form-field-type__numeric--rtl':
					Liferay.Language.direction[editingLanguageId] === 'rtl',
			})}
			disabled={readOnly}
			id={id ?? name}
			name={`${name}${inputMask ? '_masked' : ''}`}
			onBlur={onBlur}
			onChange={handleChange}
			onFocus={onFocus}
			placeholder={inputValue.placeholder}
			type="text"
			value={inputValue.masked}
		/>
	);

	return (
		<FieldBase
			{...otherProps}
			id={id}
			localizedValue={localizedValue}
			name={name}
			readOnly={readOnly}
		>
			{inputMask && append && dataType === 'double' ? (
				<ClayInput.Group>
					{appendType === 'prefix' && (
						<ClayInput.GroupItem prepend shrink>
							<ClayInput.GroupText>{append}</ClayInput.GroupText>
						</ClayInput.GroupItem>
					)}

					<ClayInput.GroupItem prepend>{input}</ClayInput.GroupItem>

					{appendType === 'suffix' && (
						<ClayInput.GroupItem append shrink>
							<ClayInput.GroupText>{append}</ClayInput.GroupText>
						</ClayInput.GroupItem>
					)}
				</ClayInput.Group>
			) : (
				input
			)}

			{inputMask && (
				<input name={name} type="hidden" value={inputValue.raw} />
			)}
		</FieldBase>
	);
};

export {Numeric};
export default withConfirmationField(Numeric);

interface IMaskedNumber {
	masked: string;
	placeholder?: string;
	raw: string;
}

interface INumberMaskConfig {
	allowDecimal?: boolean;
	allowLeadingZeroes: boolean;
	allowNegative: boolean;
	decimalLimit?: number | null;
	decimalSymbol?: string;
	includeThousandsSeparator: boolean;
	prefix?: string;
	suffix?: string;
	thousandsSeparatorSymbol?: string | null;
}

interface IProps {
	append: string;
	appendType: 'prefix' | 'suffix';
	dataType: NumericDataType;
	decimalPlaces: number;
	defaultLanguageId: Locale;
	id: string;
	inputMask?: boolean;
	inputMaskFormat?: string;
	localizedSymbols?: LocalizedValue<ISymbols>;
	localizedValue?: LocalizedValue<string>;
	name: string;
	onBlur: FocusEventHandler<HTMLInputElement>;
	onChange: FieldChangeEventHandler<String>;
	onFocus: FocusEventHandler<HTMLInputElement>;
	placeholder?: string;
	predefinedValue?: string;
	readOnly: boolean;
	settingsContext?: any;
	symbols: ISymbols;
	value?: string;
}

type NumericDataType = 'integer' | 'double';
