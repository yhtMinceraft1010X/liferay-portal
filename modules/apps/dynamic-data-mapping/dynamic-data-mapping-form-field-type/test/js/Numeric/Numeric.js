/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import '@testing-library/jest-dom/extend-expect';
import {cleanup, render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {FormProvider} from 'data-engine-js-components-web';
import React from 'react';

import Numeric from '../../../src/main/resources/META-INF/resources/Numeric/Numeric';

const globalLanguageDirection = Liferay.Language.direction;

describe('Field Numeric', () => {
	beforeAll(() => {
		Liferay.Language.direction = {en_US: 'rtl'};
	});

	afterAll(() => {
		Liferay.Language.direction = globalLanguageDirection;
	});

	afterEach(cleanup);

	it('renders the default markup', () => {
		const {container} = render(<Numeric />);

		expect(container).toMatchSnapshot();
	});

	it('has a name', () => {
		const {container} = render(<Numeric name="numericField" />);

		const input = container.querySelector('input');

		expect(input).toHaveAttribute('name', 'numericField');
	});

	it('enables input whenever readOnly is omitted', () => {
		const {container} = render(<Numeric />);

		const input = container.querySelector('input');

		expect(input).toBeEnabled();
	});

	it('disables input whenever readOnly is set', () => {
		const {container} = render(<Numeric readOnly />);

		const input = container.querySelector('input');

		expect(input).toBeDisabled();
	});

	it('has a helptext', () => {
		const {getAllByText} = render(<Numeric tip="Type something" />);

		expect(getAllByText('Type something')[0]).toBeInTheDocument();
	});

	it('has an id', () => {
		const {container} = render(<Numeric id="ID" />);

		const input = container.querySelector('input');

		expect(input).toHaveAttribute('id', 'ID');
	});

	it('has a label', () => {
		const {getAllByText} = render(<Numeric label="label" />);

		const allByText = getAllByText(/label/);
		expect(allByText).toHaveLength(2);
		expect(allByText[0]).toBeInTheDocument();
		expect(allByText[1]).toBeInTheDocument();
	});

	it('has a placeholder', () => {
		const {container} = render(<Numeric placeholder="Placeholder" />);

		const input = container.querySelector('input');

		expect(input).toHaveAttribute('placeholder', 'Placeholder');
	});

	it('is required', () => {
		const {getByText} = render(<Numeric required />);

		expect(getByText(/required/)).toBeInTheDocument();
	});

	it('renders Label if showLabel is true', () => {
		const {getAllByText} = render(
			<Numeric label="Numeric Field" showLabel />
		);

		const allByText = getAllByText(/Numeric Field/);
		expect(allByText).toHaveLength(2);
		expect(allByText[0]).toHaveClass('ddm-label');
		expect(allByText[1]).toHaveClass('sr-only');
	});

	it('has a value', () => {
		const {container} = render(<Numeric value="123" />);

		const input = container.querySelector('input');

		expect(input).toHaveValue('123');
	});

	it('fills with an input number', () => {
		const onChange = jest.fn();
		const {container} = render(<Numeric onChange={onChange} />);

		const input = container.querySelector('input');
		userEvent.type(input, '2');

		expect(onChange).toHaveBeenCalledWith({target: {value: '2'}});
	});

	it('changes the mask type', () => {
		const {container} = render(<Numeric dataType="double" value="22.22" />);

		expect(container.querySelector('input').value).toBe('22.22');
	});

	it('filters the non numeric characters when set to integer', () => {
		const onChange = jest.fn();
		const {container} = render(<Numeric onChange={onChange} />);

		const input = container.querySelector('input');
		userEvent.type(input, '3.0');

		expect(onChange).toHaveBeenLastCalledWith({
			target: {value: '30'},
		});
	});

	it('check field value is the same without decimal symbol when fieldType is integer but it receives a double', () => {
		const {container} = render(<Numeric value="3.8" />);

		const input = container.querySelector('input');

		expect(input.value).toBe('38');
	});

	it('remove decimal symbol from value when changing from decimal to integer when symbol of language is comma', () => {
		const {container} = render(
			<Numeric
				dataType="integer"
				symbols={{decimalSymbol: ','}}
				value="22,82"
			/>
		);

		expect(container.querySelector('input').value).toBe('2282');
	});

	it('updates decimal symbol using the current value of symbols', () => {
		const {container} = render(
			<Numeric
				dataType="double"
				symbols={{decimalSymbol: ','}}
				value="-1.2"
			/>
		);

		expect(container.querySelector('input').value).toBe('-1,2');
	});

	it('updates decimal symbol using the localizedSymbols based on current editing language', () => {
		const {container} = render(
			<FormProvider initialState={{editingLanguageId: 'pt_BR'}}>
				<Numeric
					dataType="double"
					localizedSymbols={{
						en_US: {
							decimalSymbol: '.',
						},
						pt_BR: {
							decimalSymbol: ',',
						},
					}}
					value="1.2"
				/>
			</FormProvider>
		);

		expect(container.querySelector('input').value).toBe('1,2');
	});

	describe('Confirmation Field', () => {
		it('renders the confirmation field with the same data type as the original field', () => {
			render(
				<Numeric
					confirmationValue="22.82"
					dataType="double"
					name="numericField"
					requireConfirmation
				/>
			);

			const confirmationField = document.getElementById(
				'numericFieldconfirmationField'
			);

			expect(confirmationField.value).toBe('22.82');
		});

		it('remove decimal symbol of the confirmation value if the data type is Integer', () => {
			render(
				<Numeric
					confirmationValue="22.82"
					name="numericField"
					requireConfirmation
				/>
			);

			const confirmationField = document.getElementById(
				'numericFieldconfirmationField'
			);

			expect(confirmationField.value).toBe('2282');
		});
	});

	describe('Integer Input Mask toggle', () => {
		it('has an inputMaskFormat', () => {
			const {container} = render(
				<Numeric
					inputMask
					inputMaskFormat="+99 (99) 9999-9999"
					name="numericField"
					value="123456789012"
				/>
			);

			expect(container).toMatchSnapshot();
		});

		it('applies mask to value', () => {
			const {container} = render(
				<Numeric
					inputMask
					inputMaskFormat="+99 (99) 9999-9999"
					value="123456789012"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('+12 (34) 5678-9012');
		});

		it('applies mask to predefined value', () => {
			const {container} = render(
				<Numeric
					inputMask
					inputMaskFormat="+99 (99) 9999-9999"
					predefinedValue="123456789012"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('+12 (34) 5678-9012');
		});

		it('truncates values over mask digit limit', () => {
			const {container} = render(
				<Numeric
					inputMask
					inputMaskFormat="+99 (099) 9999-9999"
					value="12345678901234"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('+12 (345) 6789-0123');
		});

		it('ignores optional digits whenever input is less than mandatory', () => {
			const {container} = render(
				<Numeric
					inputMask
					inputMaskFormat="+09 (099) 9999-9999"
					value="12345"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('+1 (23) 45');
		});

		it('sends unmasked value though onChange event', () => {
			const onChange = jest.fn();
			const {container} = render(
				<Numeric
					inputMask
					inputMaskFormat="E.g +99 (99) 9999-9999"
					onChange={onChange}
				/>
			);

			const input = container.querySelector('input');

			userEvent.type(input, 'E.g +55 (81) 2121-6000');

			expect(onChange).toHaveBeenLastCalledWith({
				target: {value: '558121216000'},
			});
		});

		it('limits predefined value size according to the mask', () => {
			const {container} = render(
				<Numeric
					inputMask
					inputMaskFormat="99-99"
					name="LPS-134259"
					predefinedValue="12345"
				/>
			);

			const input = container.querySelector('input[name="LPS-134259"]');

			expect(input.value).toBe('1234');
		});

		it('allows input mask format to have only numbers', () => {
			const {container} = render(
				<Numeric inputMask inputMaskFormat={99} value="1234" />
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('12');
		});

		/**
		 * This test was skipped due to an issue on userEvent.type() that not
		 * allows simulate backspace key pressing (with the current
		 * @testing-library/use-event)
		 */
		xit('it allows to delete non numeric characters from mask', () => {
			const {container} = render(
				<Numeric
					inputMask
					inputMaskFormat="99-99"
					onChange={() => {}}
					predefinedValue="12"
				/>
			);

			const input = container.querySelector('input');

			userEvent.click(input);
			userEvent.type(input, '{backspace}');

			expect(input.value).toBe('1');
		});
	});

	describe('Decimal Input Mask toggle', () => {
		it('renders a suffix', () => {
			const {container, getByText} = render(
				<Numeric
					append="$"
					appendType="suffix"
					dataType="double"
					inputMask
					name="numericField"
					value="123"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('123');
			expect(getByText('$')).toHaveClass('input-group-text');
		});

		it('renders a prefix', () => {
			const {container, getByText} = render(
				<Numeric
					append="$"
					appendType="prefix"
					dataType="double"
					inputMask
					name="numericField"
					value="123"
				/>
			);

			const input = container.querySelector('input');

			expect(getByText('$')).toHaveClass('input-group-text');
			expect(input.value).toBe('123');
		});

		it('renders the thousand separator', () => {
			const {container} = render(
				<Numeric
					dataType="double"
					inputMask
					name="numericField"
					symbols={{decimalSymbol: '.', thousandsSeparator: ','}}
					value="1234"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('1,234');
		});

		it('hides the thousand separator if it is set to `none`', () => {
			const {container} = render(
				<Numeric
					dataType="double"
					inputMask
					name="numericField"
					symbols={{decimalSymbol: '.', thousandsSeparator: 'none'}}
					value="1234"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('1234');
		});

		it('allows user to input a decimal separator', () => {
			const onChange = jest.fn();
			const {container} = render(
				<Numeric
					dataType="double"
					inputMask
					name="numericField"
					onChange={onChange}
					symbols={{decimalSymbol: ','}}
				/>
			);

			const input = container.querySelector('input');

			userEvent.type(input, '1,234');

			expect(onChange).toHaveBeenLastCalledWith({
				target: {value: '1,23'},
			});
		});

		it('generates a placeholder', () => {
			const {container} = render(
				<Numeric
					dataType="double"
					decimalPlaces="2"
					inputMask
					name="numericField"
					symbols={{decimalSymbol: ','}}
				/>
			);

			const input = container.querySelector('input');

			expect(input).toHaveAttribute('placeholder', '0,00');
		});

		it('allows user to input only the decimal quantity defined by decimal places field', () => {
			const onChange = jest.fn();
			const {container} = render(
				<Numeric
					dataType="double"
					decimalPlaces={3}
					inputMask
					name="numericField"
					onChange={onChange}
					symbols={{decimalSymbol: ','}}
				/>
			);

			const input = container.querySelector('input');

			userEvent.type(input, '1,2345678');

			expect(onChange).toHaveBeenLastCalledWith({
				target: {value: '1,234'},
			});
		});

		/**
		 * LPS-136519 / LPS-136523
		 */
		it('ignores non decimal input', () => {
			const onChange = jest.fn();
			const {container} = render(
				<Numeric
					append="999"
					appendType="suffix"
					dataType="double"
					inputMask
					name="numericField"
					onChange={onChange}
					symbols={{decimalSymbol: ','}}
				/>
			);

			const input = container.querySelector('input');

			userEvent.type(input, 'a# @e');

			expect(onChange).not.toHaveBeenCalled();
		});

		/**
		 * LPS-141862
		 */

		it('does not allow typing zeroes not followed by decimal symbol', () => {
			const onChange = jest.fn();
			const {container} = render(
				<Numeric
					dataType="double"
					decimalPlaces={3}
					inputMask
					name="numericField"
					onChange={onChange}
					symbols={{decimalSymbol: ','}}
				/>
			);

			const input = container.querySelector('input');

			userEvent.type(input, '0083,5');

			expect(onChange).toHaveBeenLastCalledWith({
				target: {value: '83,5'},
			});
		});

		it('does not allow typing sequence of zeroes', () => {
			const onChange = jest.fn();
			const {container} = render(
				<Numeric
					dataType="double"
					decimalPlaces={3}
					inputMask
					name="numericField"
					onChange={onChange}
					symbols={{decimalSymbol: ','}}
				/>
			);

			const input = container.querySelector('input');

			userEvent.type(input, '00,083');

			expect(onChange).toHaveBeenLastCalledWith({
				target: {value: '0,083'},
			});
		});
	});
});
