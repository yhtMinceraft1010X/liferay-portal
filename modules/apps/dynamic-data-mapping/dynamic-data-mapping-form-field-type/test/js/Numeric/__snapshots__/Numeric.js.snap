// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Numeric Integer Input Mask toggle has an inputMaskFormat 1`] = `
<div>
  <div
    aria-labelledby="numericField_fieldDetails"
    class="form-group hide"
    data-field-name="numericField"
    tabindex="0"
  >
    <input
      class="form-control ddm-form-field-type__numeric--rtl"
      id="numericField"
      name="numericField_masked"
      placeholder="+__ (__) ____-____"
      type="text"
      value="+12 (34) 5678-9012"
    />
    <input
      name="numericField"
      type="hidden"
      value="123456789012"
    />
    <input
      name="numericField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric renders the default markup 1`] = `
<div>
  <div
    aria-labelledby="undefined_fieldDetails"
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control ddm-form-field-type__numeric--rtl"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;
